src/Host/Client/Impl/Session/RSessionProvider.cs

```diff
@@ -61,14 +74,109 @@ public class RSessionProvider : IRSessionProvider, IRHostConnector {
                 session.Dispose();
             }
 
-            BrokerConnector.Dispose();
+            _hostConnector.Dispose();
         }
 
         private void DisposeSession(Guid guid) {
             RSession session;
             _sessions.TryRemove(guid, out session);
         }
 
+        public async Task<bool> TestBrokerConnectionAsync(string name, string path) {
+            var сonnector = await CreateConnectorAsync(name, path);
+            if (сonnector == null) {
+                return false;
+            }
+
+            var callbacks = new NullRCallbacks();
+            try {
+                var rhost = await сonnector.ConnectAsync(nameof(TestBrokerConnectionAsync), callbacks);
+                var rhostRunTask = rhost.Run();
+                callbacks.SetReadConsoleInput("q()\n");
+                await rhostRunTask;
+                return true;
+            } catch (RHostDisconnectedException) {
+                return false;
+            }
+        }
+
+        public async Task<bool> TrySwitchBrokerAsync(string name, string path = null) {
+            await TaskUtilities.SwitchToBackgroundThread();
+
+            var newConnector = await CreateConnectorAsync(name, path);
+            if (newConnector == null) {
+                return false;
+            }
+
+            // Connector switching shouldn't be concurrent
+            try {
+                await _connectorSwitchLock.WaitAsync();
+
+                // First switch connector so that all new sessions are created for the new broker
+                var oldConnector = Interlocked.Exchange(ref _hostConnector, newConnector);
+                var switchingFromNull = oldConnector is NullRHostConnector;
+                if (!switchingFromNull) { 
+                    _callback.WriteConsole(Resources.RSessionProvider_StartSwitchingWorkspaceFormat.FormatInvariant(_hostConnector.Name, GetUriString(_hostConnector)));
+                }
+
+                var sessions = _sessions.Values.ToList();
+
+                if (sessions.Any()) {
+                    try {
+                        _callback.WriteConsole(Resources.RSessionProvider_StartConnectingToWorkspaceFormat.FormatInvariant(sessions.Count));
+                        await Task.WhenAll(sessions.Select(StartSwitchingBrokerAsync));
+                        _callback.WriteConsole(Resources.RSessionProvider_RestartingSessionsFormat.FormatInvariant(sessions.Count));
+                        await Task.WhenAll(sessions.Select(s => s.CompleteSwitchingBrokerAsync()));
+                    } catch (Exception) {
```

Comment:
Should I add logging directly to the `WriteConsole`?

---
In reply to: [78040235](https://github.com/Microsoft/RTVS/pull/2268#discussion_r78040235) [](ancestors = 78040235)